### PR TITLE
Loading data enhancement. jsonline

### PR DESCRIPTION
**Pull request by mooleetzi**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [X] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #3421 

## What this PR does / why we need it:

### CODE REVIEW

File: pkg/sql/plan/build_load.go [hunk 1]
```diff
@@ -86,6 +86,9 @@ func buildLoad(stmt *tree.Load, ctx CompilerContext) (*Plan, error) {
 	tableDef.Createsql = string(json_byte)
 	node1.TableDef = tableDef
 	node1.ObjRef = objRef
+	if stmt.Param.Tail.FromJsonLine {
```
Comment (ouyuanning) [inline pkg/sql/plan/build_load.go:89]: if you Marshal stmt.Param to Createsql.
Why you set this to ExtraOptions？ may be it's not necessary, and ExtraOptions also is not a good choice

Comment (mooleetzi) [reply pkg/sql/plan/build_load.go:89]: Ok，I get it. I will remove this. 
Thx.

File: test/cases/dml/load/load.test [hunk 2]
```diff
@@ -0,0 +1,4 @@
+drop table if exists t1;
+create table t1(col1 bool,col2 int,col3 varchar, col4 date,col5 datetime,col6 timestamp,col7 decimal,col8 float);
+load data infile '$resources/load_data/jsonline.jl' into table t1 from_jsonline;
```
Comment (fengttt) [inline test/cases/dml/load/load.test:3]: How do you know this infile is a jsonline file?   Don't you need some kind of format = jsonline in the sql statement?

Comment (fengttt) [reply test/cases/dml/load/load.test:3]: Actually, let's add a comment -- 
1. format = jsonline jsondata = array    -- in this case, we are almost doing the same thing as CSV, loading each line, each element in the array is a column.
2. format = jsonline jsondata = struct   -- in this calse, we load using column names.

File: test/resources/load_data/jsonline.jl [hunk 3]
```diff
@@ -0,0 +1,2 @@
+{"col1":true,"col2":1,"col3":"var","col4":"2020-09-07","col5":"2020-09-07 00:00:00","col6":"2020-09-07 00:00:00","col7":"18","col8":121.11}
+{"col1":"true","col2":"1","col3":"var","col4":"2020-09-07","col5":"2020-09-07 00:00:00","col6":"2020-09-07 00:00:00","col7":"18","col8":"121.11"}
```
Comment (fengttt) [inline test/resources/load_data/jsonline.jl:2]: Test a case that {"col1": 1, "col2": "OK"}     {"col2": "OK", "col1": 1}

File: test/cases/dml/load/load.test (hunk 2 above)
Comment (fengttt) [reply test/cases/dml/load/load.test:3]: Ah, you have from_jsonline -- do we use from_csv?   Or use format=csv?   Keep the same as csv.
I think you need more options so this from_jsonline thing is not enough.

File: pkg/sql/colexec/external/external.go [hunk 4]
```diff
@@ -647,3 +353,342 @@ func ScanFileData(param *ExternalParam, proc *process.Process) (*batch.Batch, er
...
+				if field == "true" || field == "1" {
+					cols[rowIdx] = true
+				} else if field == "false" || field == "0" {
+					cols[rowIdx] = false
+				} else {
+					return moerr.NewInternalError("the input value '%s' is not bool type for column %d", field, colIdx)
+				}
+			}
+		case types.T_int8:
+			cols := vec.Col.([]int8)
```
Comment (fengttt) [inline pkg/sql/colexec/external/external.go:385]: use MustTCol

File: pkg/sql/colexec/external/external.go [hunk 5]
```diff
@@ -647,3 +353,342 @@ func ScanFileData(param *ExternalParam, proc *process.Process) (*batch.Batch, er
 	bat.Cnt = 1
 	return bat, nil
 }
+
+func getData(bat *batch.Batch, Line []string, rowIdx int, param *ExternalParam) error {
```
Comment (fengttt) [inline pkg/sql/colexec/external/external.go:357]: For this code, how much is shared with csv parsing? 

Can we think about if we can reduce this code using generics or something?

Comment (fengttt) [reply pkg/sql/colexec/external/external.go:357]: I see, it is completely shared with csv.   Good.

Comment (fengttt) [inline pkg/sql/colexec/external/external.go:357]: OK, so this getData, is rather expensive.   Is it possible to vectorvise it?   That is, for each batch, each column, do one switch type, then finish this column.

**Review by fengttt — COMMENTED** [review]
This change need a performanc test.

Write a script to convert TPCH 1G to jsonline format and load -- let's see how much faster/slower than CSV.

File: test/cases/dml/load/load.test (hunk 2 above)
Comment (mooleetzi) [reply test/cases/dml/load/load.test:3]: There is no from_csv or fmt=csv now. I think your thoughts are pretty good. Thx for your advice.

File: test/resources/load_data/jsonline.jl (hunk 3 above)
Comment (mooleetzi) [reply test/resources/load_data/jsonline.jl:2]: OK

File: pkg/sql/colexec/external/external.go (hunk 4 above)
Comment (mooleetzi) [reply pkg/sql/colexec/external/external.go:385]: OK

File: pkg/sql/colexec/external/external.go (hunk 5 above)
Comment (mooleetzi) [reply pkg/sql/colexec/external/external.go:357]: I will try

**Comment by mooleetzi** [comment]
> This change need a performanc test.
> 
> Write a script to convert TPCH 1G to jsonline format and load -- let's see how much faster/slower than CSV.

before vectorize `getData` function, load from 1GB csv cost 37.46 second, from jsonline array I test twice,  first time is 52.87seconds, second time is 59.79 seconds.
截屏2022-09-26 19 23 39
截屏2022-09-26 19 24 02
截屏2022-09-26 19 24 14

**Review by aressu1985 — APPROVED** [review]

**Review by fengttt — APPROVED** [review]

File: pkg/sql/plan/function/builtin/multi/external.go [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+
+import (
+	"github.com/matrixorigin/matrixone/pkg/container/nulls"
+	"github.com/matrixorigin/matrixone/pkg/container/types"
+	"github.com/matrixorigin/matrixone/pkg/container/vector"
+	"github.com/matrixorigin/matrixone/pkg/vectorize/external"
+	"github.com/matrixorigin/matrixone/pkg/vm/process"
+)
+
+type tp1 interface {
```
Comment (ouyuanning) [inline pkg/sql/plan/function/builtin/multi/external.go:25]: it's better to rename : tp1 -> tpNumber;   tp2 -> tpTime;   tp3 -> tpDecimal
as the same you can rename:
ParseCol1-> ParseNumber ,   ParseCol2-> ParseTime,   ParseCol3->ParseDecimal

Comment (mooleetzi) [reply pkg/sql/plan/function/builtin/multi/external.go:25]: Fine，I will rename them.

**Review by ouyuanning — APPROVED** [review]

**Review by iamlinjunhong — APPROVED** [review]

**Review by nnsgmsone — APPROVED** [review]

**Review by aunjgr — APPROVED** [review]

**Pull request merged**